### DISCUSSION TITLE
Linking neural-fortran to WAVEWATCH III

### DISCUSSION THREAD

**Discussion by EmilioEchevarria** [modern-fortran/neural-fortran#161] [Q&A] [answered]
Hello again, 

I've been trying to compile the WAVEWATCH III model and call some of the neural-fortran functionalities from WW3, I imagine it should be possible, right? I know this will be very specific to WW3 but I guess it could be helpful for others trying to use neural-fortran in their own code? (like [here](https://fortran-lang.discourse.group/t/using-a-trained-neural-network-model-in-a-fortran-code/5432) for example)

I use gfortran as the compiler for both. I built neural-fortran with CMake, and I can run the given examples. And I can compile WW3 with gfortran as well. 

Now in one of the WW3 subroutines in w3wavemd.F90 I add:
```
USE nf, only: network

TYPE(network) :: cnn_net
CHARACTER(LEN=200) :: keras_cnn_path = '/path/to/cnn/model.h5'
!
cnn_net = network(TRIM(keras_cnn_path))
!
call cnn_net % print_info()
```

And I try to re-compile ww3_shel. I'm using these options: 
```
-g -O3 -fno-frontend-optimize -ffree-line-length-none -fallow-argument-mismatch -L/path/to/hdf5/lib -I/path/to/hdf5/include/ -lhdf5 -lhdf5_fortran -lhdf5_hl -lhdf5hl_fortran -L/path/to/neural-fortran/build/lib -lneural -lfunctional -lh5fortran -ljsonfortran -I/path/to/neural-fortran/build/include
```

And I'm getting this error at linking time:
```
/usr/bin/ld: w3wavemd.o: in function `__w3wavemd_MOD_w3gath':
/path/to/WW3/model/src/w3wavemd.F90:3058: undefined reference to `__nf_network_MOD_network_from_keras'
/usr/bin/ld: /path/to/WW3/model/src/w3wavemd.F90:3060: undefined reference to `__nf_network_MOD_print_info'
/usr/bin/ld: /path/to/WW3/model/src/w3wavemd.F90:3246: undefined reference to `__nf_base_layer_MOD___vtab_nf
_base_layer_Base_layer'
/usr/bin/ld: w3wavemd.o: in function `.LC4':
w3wavemd.F90:(.rodata.cst8+0x8): undefined reference to `__nf_network_MOD___vtab_nf_network_Network'
collect2: error: ld returned 1 exit status
```

Any help/comment will be greatly appreciated!

Cheers,
Emilio

**Comment by EmilioEchevarria** [comment, marked as answer]
Sorry, writing down my problem actually helped me to fix it... Posting here in case it helps anybody: I had to point to the `neural-fortran` .f90 files in the WW3 link file. I added this to the link file in /model/bin:

```
nf_opt='/path/to/neural-fortran/src/*.f90 /path/to/neural-fortran/src/nf/*.f90 /path/to/neural-fortran/src/nf/io/*.f90'

opt="-o $prog $nf_opt ...."
```

And that worked. Now I can load a neural network and use it in WW3, I just need to figure out what is the best place for that in the WW3 source code... Thanks again for such a great library :)

**Reply by milancurcic** [reply]
Great to hear it, thanks for posting the solution. (I'm a wave modeler as well)
